autoload/go/def.vim

```diff
@@ -114,11 +135,122 @@ function! s:godefJump(out, mode)
 			endif
 		endif
 
-		" jump to file now
-		sil ll 1
-		normal zz
+        " Remove anything newer than the current position, just like basic
+        " vim tag support
+        if w:go_stack_level == 0
+            let w:go_stack = []
+        else
+            let w:go_stack = w:go_stack[0:w:go_stack_level-1]
+        endif
+
+		" increment the stack counter
+		let w:go_stack_level = w:go_stack_level + 1
 
-		let &switchbuf = old_switchbuf
+		" push it on to the jumpstack
+        call add(w:go_stack,
+            \{'line': line("."), 'col': col("."),
+            \'file': expand('%:p'), 'ident': a:out[1]})
+
+		" jump to file now
+        call s:goToFileLocation(location)
 	end
-	let &errorformat = old_errorformat
+endfunction
+
+function! go#def#StackPrint()
+    if len(w:go_stack) == 0
+        echohl ErrorMsg
+        echo "godef stack empty"
+        echohl None
+        return
+    endif
+    let i = 0
+    while i < len(w:go_stack)
+        let entry = w:go_stack[i]
+        if i == w:go_stack_level
+            echon "> "
+        else
+            echon "  "
+        endif
+        echon i+1 . " "
+        echohl Directory
+        echon entry["file"]
+        echohl None
+        echon "|"
+        echohl LineNr
+        echon entry["line"] . " col " . entry["col"]
+        echohl None
+        echon "|"
+        echon entry["ident"] . "\n"
+        let i += 1
+    endwhile
+    if w:go_stack_level == i
+        echo ">"
+    endif
+endfunction
```

Comment:
I think this is the wrong approach. Check out https://github.com/fatih/vim-go/blob/master/autoload/go/ui.vim#L4 how we create a custom buffer with a custom filetype and just append the content into it. It's more scalable and maintainable.